### PR TITLE
[GAT-6] Add permissions tests.

### PR DESCRIPTION
**Pull request by NigelCunningham**
This branch adds checks for permissions modified by the govcms_permissions module. It adds new Features named checkboxFromXPathChecked and checkboxFromXPathNotChecked for one of the permissions where the test used elsewhere looked at the wrong control.

### CODE REVIEW

**Comment by barbun** [comment]
hey @NigelCunningham the test looks fine but there is a better way to which used in lightning I pointed you earlier today.
https://github.com/acquia/lightning/blob/8.x-2.x/tests/features/bootstrap/lightning.behat.inc

See assertPermissions and assertNoPermissions functions.
And an example would be:

Given I am logged in as a user with the "administer permissions" permission
    When I visit "/admin/people/permissions"
    Then the page_reviewer role should have permission to:
      """
      view moderation states
      """
    And the landing_page_reviewer role should have permission to:
      """
      view moderation states
      """

**Comment by NigelCunningham** [comment]
Test this

File: build/tests/behat/bootstrap/FeatureContext.php [hunk 1]
```diff
@@ -262,6 +262,117 @@ public function iShouldBeAbleToCancelTheAccount($username) {
...
+  private function roleToRid($rid) {
+    if (is_numeric($rid)) {
+      return $rid;
+    }
+
+    return array_search($rid, user_roles());
+  }
+
+  /**
+   * Asserts that a role as a set of permissions.
```
Comment (barbun) [inline build/tests/behat/bootstrap/FeatureContext.php]: I assume that is a typo. Needs to be "...a role has a set of permission"

File: build/tests/behat/features/permissions.feature [hunk 2]
```diff
@@ -0,0 +1,884 @@
+Feature: Permissions
+  Check that permissions are as expected.
+
+  @api @role @drupal @wip
```
Comment (barbun) [inline build/tests/behat/features/permissions.feature]: Can @wip tag be removed?

File: build/tests/behat/features/permissions.feature [hunk 3]
```diff
@@ -0,0 +1,884 @@
+Feature: Permissions
+  Check that permissions are as expected.
```
Comment (barbun) [inline build/tests/behat/features/permissions.feature]: I don't mean to be super strict on this but I'd like the descriptions to be a bit more clear. E.g. "Check that the default govCMS permissions are as expected"

File: build/tests/behat/features/permissions.feature [hunk 4]
```diff
@@ -0,0 +1,884 @@
...
+      view moderation messages
+      use workbench_moderation my drafts tab
+      use workbench_moderation needs review tab
+      moderate content from draft to needs_review
+      moderate content from needs_review to draft
+      moderate content from needs_review to published
+      administer xmlsitemap
+      """
+
+  @api @role @wip
```
Comment (barbun) [inline build/tests/behat/features/permissions.feature]: Same deal with the a test @wip tag

File: build/tests/behat/features/permissions.feature [hunk 5]
```diff
@@ -0,0 +1,884 @@
...
+  Scenario: Check authenticated user permissions
+    Given I am logged in as a user named "betty_perms" with the "administrator" role that doesn't force password change
```
Comment (barbun) [inline build/tests/behat/features/permissions.feature]: Can this be "Given I am logged in as a user with the "administer permissions" permission and don't need a password change" instead?

File: build/tests/behat/features/permissions.feature [hunk 6]
```diff
@@ -0,0 +1,884 @@
...
+      view moderation messages
+      use workbench_moderation my drafts tab
+      use workbench_moderation needs review tab
+      moderate content from draft to needs_review
+      moderate content from needs_review to draft
+      moderate content from needs_review to published
+      administer xmlsitemap
+      """
+
+  @api @role @wip
```
Comment (barbun) [inline build/tests/behat/features/permissions.feature]: Again, @wip

File: build/tests/behat/features/permissions.feature [hunk 7]
```diff
@@ -0,0 +1,884 @@
...
+  Scenario: Check Content editor user permissions
+    Given I am logged in as a user named "christine_perms" with the "administrator" role that doesn't force password change
```
Comment (barbun) [inline build/tests/behat/features/permissions.feature]: Consider "Given I am logged in as a user with the "administer permissions" permission and don't need a password change" for consistency. We aren't testing admin role.

File: build/tests/behat/features/permissions.feature [hunk 8]
```diff
@@ -0,0 +1,884 @@
...
+      edit all webform submissions
+      delete all webform submissions
+      administer workbench moderation
+      set webform_clear times
+      bypass workbench moderation
+      moderate content from needs_review to published
+      administer xmlsitemap
+      """
+
+  @api @role @wip
```
Comment (barbun) [inline build/tests/behat/features/permissions.feature]: Again, testing tag.

File: build/tests/behat/features/permissions.feature [hunk 9]
```diff
@@ -0,0 +1,884 @@
...
+  Scenario: Check Content approver user permissions
+    Given I am logged in as a user named "daniel_perms" with the "administrator" role that doesn't force password change
```
Comment (barbun) [inline build/tests/behat/features/permissions.feature]: Same deal:
Given I am logged in as a user with the "administer permissions" permission and don't need a password change

File: build/tests/behat/features/permissions.feature [hunk 10]
```diff
@@ -0,0 +1,884 @@
...
+      access all views
+      edit all webform submissions
+      delete all webform submissions
+      set webform_clear times
+      administer workbench moderation
+      bypass workbench moderation
+      administer xmlsitemap
+      """
+
+  @api @role @wip
```
Comment (barbun) [inline build/tests/behat/features/permissions.feature]: wip

File: build/tests/behat/features/permissions.feature [hunk 11]
```diff
@@ -0,0 +1,884 @@
...
+  Scenario: Check Site Builder user permissions
+    Given I am logged in as a user named "eric_perms" with the "administrator" role that doesn't force password change
```
Comment (barbun) [inline build/tests/behat/features/permissions.feature]: Given I am logged in as a user with the "administer permissions" permission and don't need a password change

File: build/tests/behat/features/permissions.feature [hunk 12]
```diff
@@ -0,0 +1,884 @@
...
+      delete terms in 2
+      schedule publishing of nodes
+      edit own webform submissions
+      delete own webform submissions
+      moderate content from draft to needs_review
+      moderate content from needs_review to draft
+      moderate content from needs_review to published
+      """
+
+  @api @role @wip
```
Comment (barbun) [inline build/tests/behat/features/permissions.feature]: wip

File: build/tests/behat/features/permissions.feature [hunk 13]
```diff
@@ -0,0 +1,884 @@
...
+  Scenario: Check Site editor user permissions
+    Given I am logged in as a user named "fred_perms" with the "administrator" role that doesn't force password change
```
Comment (barbun) [inline build/tests/behat/features/permissions.feature]: Given I am logged in as a user with the "administer permissions" permission and don't need a password change

File: build/tests/behat/features/permissions.feature [hunk 14]
```diff
@@ -0,0 +1,884 @@
...
+      administer content types
+      edit own webform content
+      delete own webform content
+      assign all roles
+      schedule publishing of nodes
+      edit terms in 2
+      delete terms in 2
+  """
+
+  @api @role @wip
```
Comment (barbun) [inline build/tests/behat/features/permissions.feature]: wip

File: build/tests/behat/features/permissions.feature [hunk 15]
```diff
@@ -0,0 +1,884 @@
...
+  Scenario: Check Administrator user permissions
+    Given I am logged in as a user named "gerry_perms" with the "administrator" role that doesn't force password change
```
Comment (barbun) [inline build/tests/behat/features/permissions.feature]: Given I am logged in as a user with the "administer permissions" permission and don't need a password change

File: build/tests/behat/bootstrap/FeatureContext.php [hunk 16]
```diff
@@ -300,4 +411,36 @@ protected function selectUserVBOCheckbox($username) {
     }
   }
 
+  /**
+   * Checks that specific checkbox is checked.
+   *
+   * @Then /^checkbox from xpath "(?P<field>[^"]*)" should be checked$/
+   */
+  public function checkboxFromXPathChecked($field, TraversableElement $container = NULL) {
```
Comment (barbun) [inline build/tests/behat/bootstrap/FeatureContext.php]: Where is this being used?

File: build/tests/behat/bootstrap/FeatureContext.php [hunk 17]
```diff
@@ -300,4 +411,36 @@ protected function selectUserVBOCheckbox($username) {
...
+      throw new \InvalidArgumentException(sprintf('Checkbox "%s" is not checked, but it should be.', $field));
+    }
+  }
+
+  /**
+   * Checks that specific checkbox is unchecked.
+   *
+   * @Then /^checkbox from xpath "(?P<field>[^"]*)" should not be checked$/
+   */
+  public function checkboxFromXPathNotChecked($field, TraversableElement $container = NULL) {
```
Comment (barbun) [inline build/tests/behat/bootstrap/FeatureContext.php]: Where is this function being used?

File: build/tests/behat/bootstrap/FeatureContext.php (hunk 16 above)
Comment (NigelCunningham) [inline build/tests/behat/bootstrap/FeatureContext.php]: With the changes you requested, it's not being used anymore. Removed.

File: build/tests/behat/bootstrap/FeatureContext.php (hunk 17 above)
Comment (NigelCunningham) [inline build/tests/behat/bootstrap/FeatureContext.php]: With the changes you requested, it's not being used anymore. Removed.

**Pull request merged**